src/libraries/Microsoft.PowerFx.Connectors/Public/ConnectorType.cs

```diff
@@ -271,24 +271,25 @@ internal void SetRelationship(SqlRelationship relationship)
             ForeignKey = relationship.ReferencedColumnName;
         }
 
-        internal void AddTabularDataSource(ICdpTableResolver tableResolver, IList<ReferencedEntity> referencedEntities, List<SqlRelationship> sqlRelationships, DName name, string datasetName, ConnectorType connectorType, ServiceCapabilities serviceCapabilities, bool isReadOnly, BidirectionalDictionary<string, string> displayNameMapping = null)
+        internal void AddTabularDataSource(ICdpTableResolver tableResolver, IList<ReferencedEntity> referencedEntities, List<SqlRelationship> sqlRelationships, DName name, string datasetName, ServiceCapabilities serviceCapabilities, bool isReadOnly, BidirectionalDictionary<string, string> displayNameMapping = null)
         {
             if (FormulaType is not RecordType)
             {
                 throw new PowerFxConnectorException("Invalid FormulaType");
             }
 
-            // $$$ Hack to enable IExternalTabularDataSource, will be removed later
-#pragma warning disable CS0618 // Type or member is obsolete
-            if (tableResolver.GenerateADS)
-            {
-                HashSet<IExternalTabularDataSource> dataSource = new HashSet<IExternalTabularDataSource>() { new ExternalCdpDataSource(name, datasetName, serviceCapabilities, isReadOnly, displayNameMapping) };
-                DType newDType = DType.CreateDTypeWithConnectedDataSourceInfoMetadata(FormulaType._type, dataSource, null);
-                FormulaType = new KnownRecordType(newDType);
-            }
-#pragma warning restore CS0618 // Type or member is obsolete
+            HashSet<IExternalTabularDataSource> dataSource = new HashSet<IExternalTabularDataSource>() { new ExternalCdpDataSource(this, name, datasetName, serviceCapabilities, isReadOnly, displayNameMapping) };
+            DType newDType = DType.CreateDTypeWithConnectedDataSourceInfoMetadata(FormulaType._type, dataSource, null);
+            FormulaType = new KnownRecordType(newDType);
+
+            FormulaType = new CdpRecordType(this, FormulaType._type, tableResolver, referencedEntities, sqlRelationships);
+        }
 
-            FormulaType = new CdpRecordType(connectorType, FormulaType._type, tableResolver, referencedEntities, sqlRelationships);
+        internal static FormulaType GetRecordTypeWithADS(RecordType recordType, IList<ReferencedEntity> referencedEntities, List<SqlRelationship> sqlRelationships, DName name, string datasetName, ServiceCapabilities serviceCapabilities, bool isReadOnly, BidirectionalDictionary<string, string> displayNameMapping = null)
```

Comment:
>GetRecordTypeWithADS [](http://example.com/codeflow?start=36&length=20)

Used for testing only

Reply:
If its only used in testing, can we move this as an extension method in that test assembly, having test specific code with prod code seems risky, someone may end up using it unknowingly in future.

Reply:
Not only for testing, I've reworked the code a bit to make it cleaner